docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+
+Real Kubernetes clusters have a variety of volumes which differ widely in
+size, iops performance, retention policy, and other characteristics.
+Administrators need a way to dynamically provision volumes of these different
+types to automatically meet user demand.
+
+A new mechanism called 'storage classes' is proposed to provide this
+capability.  Storage classes build upon the label selection feature added in
+Kubernetes 1.3.
+
+## Motivation
+
+In Kubernetes 1.2, an alpha form of limited dynamic provisioning was added
+that allows a single volume type to be provisioned in clouds that offer
+special volume types.
+
+In Kubernetes 1.3, a label selector was added to persistent volume claims to
+allow administrators to create a taxonomy of volumes based on the
+characteristics important to them, and to allow users to make claims on those
+volumes based on those characteristics.  This allows flexibility when claiming
+existing volumes; the same flexibility is needed when dynamically provisioning
+volumes.
+
+After gaining experience with dynamic provisioning after the 1.2 release, we
+want to create a more flexible feature that allows configuration of how
+different storage classes are provisioned and supports provisioning multiple
+types of volumes within a single cloud.
+
+## Constraints and Assumptions
+
+The proposed design should:
+
+1.  Build upon the label selection capabilities added in Kubernetes 1.3
+2.  Not interfere with existing semantics of manually-created volumes
+3.  Not require users to know or understand the differences between
+    volumes (ie, Kubernetes should not dictate any particular set of
+    characteristics to administrators to think in terms of)
+
+This proposal will **not** deal with out-of-tree provisioners, but we should
+understand that they are desired by the community, and so we should be careful
+not to make them impossible to implement in the future.
+
+## Use Cases
+
+1.  As an administrator, I want to be able to provision multiple configurations
+    of various volume types within a single cluster, and to parameterize and
+    configure the particulars of what is provisioned
+
+### Use Case: Define how volumes are provisioned
+
+Currently, there is an alpha version of dynamically provisioning, which allows
+provisioning of a **certain, single** volume type in clouds which offer
+natively supported persistent volumes.  For example, in clusters running on
+AWS, it is possible to provision AWS EBS volumes dynamically.  Real clusters
+require additional flexibility, such as the ability to differentiate different
+classes of storage to be provisioned, and the ability to provision volumes of
+different types.
+
+Using the new label selection feature for persistent volumes will allow
+storage classes to function as an extension of the existing API.  Storage
+classes should be matched with persistent volume claims using the same
+selector field as volumes, if no volumes are available that match a claim's
```

Comment:
"if no volumes are available that match a claim's selector" - this changes current behavior. In 1.2 (and current master) we **always** provision volumes when there is `volume.alpha.kubernetes.io/storage-class` annotation present on PVC, even if there is a matching PV available.